internal/service/invoice.go

```diff
@@ -1318,3 +1345,64 @@ func (s *invoiceService) RecalculateInvoice(ctx context.Context, id string, fina
 	// Return updated invoice
 	return s.GetInvoice(ctx, id)
 }
+
+// handleCouponOverrides handles coupon overrides for an invoice
+func (s *invoiceService) applyCouponsToInvoice(ctx context.Context, inv *invoice.Invoice, req dto.CreateInvoiceRequest) error {
+	// Use coupon service to prepare and apply coupons
+	couponService := NewCouponService(s.ServiceParams)
+
+	// Prepare coupons for the invoice
+	couponsWithAssociations, err := couponService.PrepareCouponsForInvoice(ctx, req)
+	if err != nil {
+		return err
+	}
+
+	// Apply coupons to the invoice
+	couponResult, err := couponService.ApplyCouponsOnInvoice(ctx, inv, couponsWithAssociations)
+	if err != nil {
+		return err
+	}
+
+	// Update the invoice with calculated discount amounts
+	inv.TotalDiscount = couponResult.TotalDiscountAmount
+
+	// Calculate new total, ensuring it doesn't go below zero
+	originalTotal := inv.Total
+	newTotal := originalTotal.Sub(couponResult.TotalDiscountAmount)
+
+	// Ensure total doesn't go negative
+	if newTotal.LessThan(decimal.Zero) {
+		s.Logger.Warnw("discount amount exceeds invoice total, capping at zero",
+			"invoice_id", inv.ID,
+			"original_total", originalTotal,
+			"total_discount", couponResult.TotalDiscountAmount,
+			"calculated_total", newTotal)
+		newTotal = decimal.Zero
+		// Adjust the total discount to not exceed the original total
+		inv.TotalDiscount = originalTotal
+	}
+
+	inv.Total = newTotal
+
+	// Update AmountDue and AmountRemaining to reflect new total
+	inv.AmountDue = newTotal
+	inv.AmountRemaining = newTotal.Sub(inv.AmountPaid)
+
+	// Update the invoice in the database
+	if err := s.InvoiceRepo.Update(ctx, inv); err != nil {
```

Comment:
inside ent invoice repo call inv.Validate to call domain invoice validation

also make sure the discount related maths is part of the domain validation